### PR TITLE
[RFC] Rework address book from commands to crud

### PR DESCRIPTION
**Pull request by Tomanhez**

### CODE REVIEW

**Comment by mamazu** [comment]
Thanks for your pull request, the idea is great the problem I see though is that this changes the response codes and the response structure very drastically.

File: tests/Responses/Expected/address_book/validation_create_address_book_response.json [hunk 1]
```diff
@@ -1,25 +1,40 @@
...
-    "postcode": [
-      "@string@"
-    ]
-  }
+    "children": {
+      "firstName": {
+        "errors": [
+          "Please enter first name."
+        ]
+      },
```
Comment (lchrusciel) [inline tests/Responses/Expected/address_book/validation_create_address_book_response.json]: Something went wrong with the formatting in this file. Can we clean it?

**Comment by mamazu** [comment]
Well the validation message structure is different. I'd say that this is a problem indeed. If there is a way to hook into that, this would be great other than that this would cause a very big inconsistency with the current validation views, which I wouldn't really like.

File: src/Resources/config/app/config.yml [hunk 2]
```diff
@@ -10,6 +10,9 @@ jms_serializer:
             sylius-shop-api-placed-order:
                 namespace_prefix: "Sylius\\ShopApiPlugin\\View\\Order"
                 path : "@ShopApiPlugin/Resources/config/serializer/Order"
+            sylius-shop-api-sylius-core:
```
Comment (lchrusciel) [inline src/Resources/config/app/config.yml]: Should be removed

File: tests/Controller/AddressBook/AddressBookCreateAddressApiTest.php [hunk 3]
```diff
@@ -38,7 +38,7 @@ public function it_allows_user_to_add_new_address_to_address_book(): void
         }
 EOT;
 
-        $this->client->request('POST', '/shop-api/WEB_GB/address-book', [], [], self::CONTENT_TYPE_HEADER, $data);
+        $this->client->request('POST', '/shop-api/WEB_GB/address-book/', [], [], self::CONTENT_TYPE_HEADER, $data);
```
Comment (lchrusciel) [inline tests/Controller/AddressBook/AddressBookCreateAddressApiTest.php]: We should mention about this change in [UPGRADE.md](https://github.com/Sylius/ShopApiPlugin/blob/master/UPGRADE.md#upgrade-from-100-beta21-to-100-beta22)

File: tests/Responses/Expected/address_book/show_address_book_response.json [hunk 4]
```diff
@@ -3,39 +3,44 @@
...
         "firstName": "Jeanie",
         "lastName": "Metz",
+        "phoneNumber": "349713",
         "countryCode": "GB",
+        "provinceCode": "GB-SCT",
         "street": "Kupreska",
         "city": "Klingside",
-        "postcode": "33553",
-        "provinceCode": "GB-SCT",
-        "company": "Locastic",
```
Comment (lchrusciel) [inline tests/Responses/Expected/address_book/show_address_book_response.json]: Oh heck, we are losing some part of the information. Can you investigate it?

File: tests/Responses/Expected/address_book/validation_create_address_book_response.json (hunk 1 above)
Comment (lchrusciel) [reply tests/Responses/Expected/address_book/validation_create_address_book_response.json]: up

File: tests/Responses/Expected/address_book/show_address_book_response.json (hunk 4 above)
Comment (mamazu) [reply tests/Responses/Expected/address_book/show_address_book_response.json]: Maybe because it is a different serializer configuration?

File: UPGRADE.md [hunk 5]
```diff
@@ -40,9 +40,12 @@
     | `{channelCode}/verify-account`           | `verify-account`                    |
     | `{channelCode}/taxon-products/*`         | `taxon-products/*`                  |
     | `{channelCode}/taxons/*`                 | `taxons/*`                          |
+   
+* The address-book create route have been changed :
```
Comment (GSadee) [inline UPGRADE.md]:
```suggestion
* The address-book create route has been changed :
```

File: tests/Controller/AddressBook/AddressBookCreateAddressApiTest.php [hunk 6]
```diff
@@ -28,13 +28,14 @@ public function it_allows_user_to_add_new_address_to_address_book(): void
         $data =
 <<<JSON
         {
-            "firstName": "Jurica",
-            "lastName": "Separovic",
-            "phoneNumber": "091892212",
-            "city": "Split",
-            "street": "Kupreska 12",
+            "firstName": "New name",
```
Comment (GSadee) [inline tests/Controller/AddressBook/AddressBookCreateAddressApiTest.php]: Unnecessary changes

**Comment by lchrusciel** [comment]
Let's merge it and see what feedback from the community we will receive. In any case, we can always revert this change.

**Comment by lchrusciel** [comment]
Thanks, Tomasz! :1st_place_medal:

**Comment by diimpp** [comment]
Why this change was necessary? How serialization will handle case of different data format for SyliusAdminApi and SyliusShopApi? Now I need to start serialization groups hell again, only at ShopApi side.

Is validation consistent with rest of the ShopApi output as @lchrusciel asked before?

-1 for the change, I think I will keep original implementation.

**Comment by lchrusciel** [comment]
No, validation is not consistent yet. 

Thanks a lot for your opinion. I'm not insisting on this implementation, I'm ready to move back to the previous one if this will be the request from the community. Of course, I want to make a decision about it before 1.0

**Comment by diimpp** [comment]
Thank you for clarification.

Base output have changed as well as address renders customer details (and other unnecessary details through customer relationship).

To clarify my point,

Conceptually, till this change, ShopApi was "add-on" (I'm aware, that it's not really compatible with SyliusShop, but that's a technical difficulty, not intentional change), but right now it overrides parent Sylius serialization mappings for Address

```yaml
jms_serializer:
    metadata:
        directories:
            # ...
            sylius-addressing:
                namespace_prefix: "Sylius\\Component\\Addressing"
                path: "@SyliusShopApiPlugin/Resources/config/serializer/Address"
```
which makes it intentionally tamper with Sylius Core.

I think such decision for conception shift should be weighted closely and in my opinion a few boiler-plate classes around address book are not really a reason for this approach.
